### PR TITLE
TEMP PR: Group collection activeselection rewrite

### PR DESCRIPTION
**Pull request by asturur**
This PR brings in the rewrite made by @ShaMan123 for the group and collection and active selection.
There is no new feature that can be seen, but the group class now is able to self adjust the bounding box.

### canvas_grouping.mixin.js
removed addWithUpdate and removeWithUpdate with add and remove.

### collection.mixin.js
add insertAt and remove now take an array of items rather than a single item and they take a callback.
the callback replace `_onObjectAdded` and `_onObjectRemoved` functions in the classes.
The reason for this is that a classes may want to do different things when adding or removing for different reasons.
Add Remove and InsertAt are now the `withUpdate` version, and the without update does not exist anymore.
getObjects now can take more types.

Worth to note that as of now the collection add/remove are never used directly but always wrapped in another method from 
the class that inherits from collection.

### active_selection.class.js
Not much changed. 
toGroup became exitGroup ( verify if the name of the method sound correct, together with enterGroup in group )

### object.class.js
The change isn't related to the PR and sneak in for some reason, probably because in looking at UTs changes i noticed a strong dependency between `includeDefaultValues` and `stateProperties`. Depending how many UT fails removing that change it could be worth putting it in its own commit.

### static_canvas.class.js 
add the internal implementation for add, remove and insertAt followed by a possible renderAll.
The implementation just call the collection one.
Notes:
- eventually specify the arguments in the functions.
- consider wrapping useless calls ( with no objects for example ) wit ha DEV_START regex and a console warning
- remove has a `didRemove` logic, while add and insertAt don't

### misc.js 
Removed the dependency of SVG width/height in the groupSVGElements.
Before those changes group allowed to specify which was the centerPoint for the group and run special logic on it. Was an hack from when removed the PathGroup that was a specific container for parsed SVGs.
groupSVGElements can simply group thing. The specific logic that is there about trying to keep the SVG viewport was an hack and should be implemented with a clipPath or some other render logic, or could be ignored and developers need to clean the SVG from cross border objects.

### group.class.js
Complete rewrite of the class.
Added a layout property to define the behavour of the group. More layout type will be defined in following PRs or supported as subclasses, depending on what seems to make more sense.
Added interactive property, for now disabled, used to permit to interact with objects inside the group
The initialization now attach an event listener to objects for modified, changed and selected/deselected.
When those event fire, the group will trigger a layout change.
(more to come)

### Activeselection
- removed the specific tests for add/remove/insert add and add generic one. ( tests were duplicated ? )

### Canvas
Nothing ( swapped removeWIthUpdate and AddWithUpdate with remove/add and added a couple of setCoords after a change in subTargetCheck, that is completely fine. No real functional changes, but the setCoords required needs to be noted in the changelog )

### StaticCanvas
extended the test about renderOnAddRemove, nothing else changed.

### Collection
Adapted tests to the new add/remove/insert att argument signature with a callback.
Removed the constraints about chainability

### Group
removed chainability constrains for add, remove, insertAt
Change the object shape to consider the new properties
rename destroy with removeAll

## CHANGES compared to v6!

### CODE REVIEW

File: src/mixins/collection.mixin.js [hunk 1]
```diff
@@ -3,79 +3,69 @@
...
+   * @param {(object:fabric.Object) => any} [callback]
    */
-  insertAt: function (object, index, nonSplicing) {
-    var objects = this._objects;
-    if (nonSplicing) {
-      objects[index] = object;
-    }
-    else {
-      objects.splice(index, 0, object);
+  insertAt: function (objects, index, callback) {
```
Comment (asturur) [inline src/mixins/collection.mixin.js:38]: modified to take more than one object as add and remove always did.
rare use case, maybe none, but good for consistency.
callback added because group will have 2 different equivalent of `onObjectAdded`

File: src/shapes/active_selection.class.js [hunk 2]
```diff
@@ -24,58 +24,42 @@
...
+
+    /**
+     * @override
+     */
+    subTargetCheck: false,
+
+    /**
+     * @override
+     */
+    interactive: false,
```
Comment (asturur) [inline src/shapes/active_selection.class.js:40]: subTargetCheck and being able to select and modify sub objects are 2 distinct use cases.

File: src/shapes/object.class.js [hunk 3]
```diff
@@ -899,10 +899,9 @@
      * @param {Object} object
      */
     _removeDefaultValues: function(object) {
-      var prototype = fabric.util.getKlass(object.type).prototype,
-          stateProperties = prototype.stateProperties;
-      stateProperties.forEach(function(prop) {
-        if (prop === 'left' || prop === 'top') {
+      var prototype = fabric.util.getKlass(object.type).prototype;
+      Object.keys(object).forEach(function(prop) {
```
Comment (asturur) [inline src/shapes/object.class.js:903]: modified so that all properties of the exported object get compared to the prototype.

File: src/static_canvas.class.js [hunk 4]
```diff
@@ -32,7 +32,8 @@
    * @fires object:added
    * @fires object:removed
    */
-  fabric.StaticCanvas = fabric.util.createClass(fabric.CommonMethods, /** @lends fabric.StaticCanvas.prototype */ {
+  // disable-eslint-next-line max-len
+  fabric.StaticCanvas = fabric.util.createClass(fabric.CommonMethods, fabric.Collection, /** @lends fabric.StaticCanvas.prototype */ {
```
Comment (asturur) [inline src/static_canvas.class.js:36]: collection was added at the end of the class.
Since we want to override some methods we can't do that.

Comment (ShaMan123) [reply src/static_canvas.class.js:36]: correct

File: src/shapes/active_selection.class.js (hunk 2 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:40]: I can commit this to v6! if you want, thought of it too

File: src/mixins/collection.mixin.js (hunk 1 above)
Comment (ShaMan123) [reply src/mixins/collection.mixin.js:38]: yep

File: src/shapes/active_selection.class.js (hunk 2 above)
Comment (asturur) [reply src/shapes/active_selection.class.js:40]: we will need to rebase on this once we bring another piece on master. So if this work out it will go in naturally.

**Comment by asturur** [comment]
i - need - to - finish - this

**Comment by asturur** [comment]
Somehow this PR breaks svg import

**Comment by ShaMan123** [comment]
> Somehow this PR breaks svg import

https://github.com/fabricjs/fabric.js/pull/7669/files#diff-d2af815271fe68bc090b7f0a204a74bc05a4289fe5d8f05467fec056d757085e

https://github.com/fabricjs/fabric.js/pull/7669#issuecomment-1053704270

**Comment by asturur** [comment]
i remember this discussion 3 minutes after i posted the comment. I came back now determined to fix it.

**Comment by ShaMan123** [comment]
When you have time let's talk. I want to understand how you plan on moving forward.

**Comment by asturur** [comment]
We can talk monday if you want, or tomorrow. But for tomorrow i have no idea when i m free.
When the baby fall asleep in the afternoon i can try to ping you and see if you are around.

The idea anyway is to merge your work in small pieces, this is the biggest piece that i couldn't really divide in sub pieces

**Comment by asturur** [comment]
GroupSvgElement isn't broken at all, grouping works, but is the final position that differs.
I changed a bit the group initialization to make it work, i ll comment and ask you question.

**Comment by ShaMan123** [comment]
> We can talk monday if you want, or tomorrow. But for tomorrow i have no idea when i m free. When the baby fall asleep in the afternoon i can try to ping you and see if you are around.
> 
> The idea anyway is to merge your work in small pieces, this is the biggest piece that i couldn't really divide in sub pieces

Works

File: src/shapes/group.class.js [hunk 5]
```diff
@@ -15,280 +21,318 @@
...
-      options = options || {};
-      this._objects = [];
-      // if objects enclosed in a group have been grouped already,
-      // we cannot change properties of objects.
-      // Thus we need to set options to group without objects,
-      isAlreadyGrouped && this.callSuper('initialize', options);
+     * @param {boolean} [objectsRelativeToGroup] true if objects exist in group coordinate plane
+     * @return {fabric.Group} thisArg
+     */
+    initialize: function (objects, options, objectsRelativeToGroup) {
```
Comment (asturur) [inline src/shapes/group.class.js:92]: which are the cases in which `objectsRelativeToGroup` would be true, apart when loading from json/object?

**Comment by ShaMan123** [comment]
Svg as well

בתאריך שבת, 26 במרץ 2022, 23:37, מאת Andrea Bogazzi ‏<
***@***.***>:

File: src/shapes/group.class.js [hunk 6]
```diff
@@ -15,280 +21,318 @@
...
+        this.forEachObject(function (object) {
+          this.enterGroup(object, false);
+        }, this);
       }
       else {
-        this._updateObjectsACoords();
+        //  we need to preserve object's center point in relation to canvas and apply group's transform to it
+        var inv = invertTransform(this.calcTransformMatrix());
+        this.forEachObject(function (object) {
+          var center = transformPoint(object.getCenterPoint(), inv);
```
Comment (asturur) [inline src/shapes/group.class.js:108]: this line was:
```
var center = transformPoint(this.getCenterPoint(), t);
```

The center point was the one from the group, and the matrix was (inv * objectTransform)

now, if i m not wrong the objectTransform would move the the group center ( the 0,0 of this plane ) in the object center, taking in account also the object trasnform. But getCenterPoint is doing the same somehow.

I not 100% this change is equivalent, but fixes the svg import issue. Do you think is breaking something else?

Comment (asturur) [reply src/shapes/group.class.js:108]: is this supposed to take in account creating group with objects that are part of other groups and already deeply transformed?

File: src/shapes/group.class.js (hunk 5 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:92]: We shouldn't do that. Mainly because I want the layout mechanism to be easily overridden.

File: src/shapes/active_selection.class.js (hunk 2 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:40]: Why do you think it's 2 distinct use cases?

File: src/shapes/group.class.js [hunk 7]
```diff
@@ -15,280 +21,318 @@
...
+          this.enterGroup(object, false);
```
Comment (ShaMan123) [inline src/shapes/group.class.js:109]: > is this supposed to take in account creating group with objects that are part of other groups and already deeply transformed?

This is in charge of it.

File: src/shapes/group.class.js (hunk 6 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:108]: I think this was a bug.
It should be the object center that is transformed into the group's plane.
Great work!

Comment (ShaMan123) [reply src/shapes/group.class.js:108]: I have an idea for a test to write for this. Can I commit to v6!? or should I open a PR?

Comment (ShaMan123) [reply src/shapes/group.class.js:108]: I want to commit this fix to v6! as well.

File: src/util/misc.js [hunk 8]
```diff
@@ -607,31 +607,19 @@
...
           delete options.width;
           delete options.height;
         }
       }
-      object = new fabric.Group(elements, options);
-      if (typeof path !== 'undefined') {
-        object.sourcePath = path;
-      }
-      return object;
+      return new fabric.Group(elements, options);
```
Comment (ShaMan123) [inline src/util/misc.js:622]:
```suggestion
      return new fabric.Group(elements, Object.assign(options || {}, { layout: 'svg' }));
```

You missed this

File: src/shapes/group.class.js (hunk 5 above)
Comment (asturur) [reply src/shapes/group.class.js:92]: ok but isn't the layout mechanism unnecessary if you are reloading a group from object or a subclass? the properties of the layout are already saved in your json form

Comment (ShaMan123) [reply src/shapes/group.class.js:92]: In most use cases it is unnecessary.
But who knows what's the edge case.
The layout mechanism handles this case internally from `prepareInitialBoundingBox`

Comment (ShaMan123) [reply src/shapes/group.class.js:92]: I thought svg import uses `objectsRelativeToGroup` but I see it doesn't

File: src/shapes/group.class.js (hunk 6 above)
Comment (asturur) [reply src/shapes/group.class.js]: we need to polish and merge this PR, either by you with a patch or we just merge this.
Changing this piece of code break a single UTs, it has some differences for nested groups.

File: src/util/misc.js (hunk 8 above)
Comment (asturur) [reply src/util/misc.js]: So the issue with the SVG is that the group creation code is perfectly able to mantain position and size, but then the position was changed from the layout strategy.
Removing options entirely works, but of course it does not work for SVG that are outside the bounding box.

File: src/shapes/active_selection.class.js (hunk 2 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:40]: Alright I think I understand.
e.g. ActiveSelection should not be interactive be should have subTargetCheck set to true?

File: src/shapes/group.class.js (hunk 6 above)
Comment (ShaMan123) [reply src/shapes/group.class.js]: I'll patch and merge.
What are the diffs in tests?

Comment (ShaMan123) [reply src/shapes/group.class.js]: master:
image

before your fix:
image


with your fix:
image

I'll push a visual test for this

File: src/shapes/group.class.js [hunk 9]
```diff
@@ -105,7 +105,11 @@
         //  we need to preserve object's center point in relation to canvas and apply group's transform to it
         var inv = invertTransform(this.calcTransformMatrix());
         this.forEachObject(function (object) {
-          var center = transformPoint(object.getCenterPoint(), inv);
```
Comment (ShaMan123) [inline src/shapes/group.class.js]: Why did you remove this???
It is correct

**Comment by ShaMan123** [comment]
It is becoming hard to work like this, with so many branches and conflicts

File: src/shapes/group.class.js (hunk 9 above)
Comment (asturur) [reply src/shapes/group.class.js]: i put back the original code because i have a failing UT.
I want all the test to pass and the figure out what is going on.
Now i have 2 completely unrelated UTs failing only in browser

Comment (ShaMan123) [reply src/shapes/group.class.js]: > master: image
> 
> before your fix: image
> 
> with your fix: image
> 
> I'll push a visual test for this

Comment (ShaMan123) [reply src/shapes/group.class.js]: @asturur the UT is wrong

File: src/shapes/group.class.js (hunk 6 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:108]: This test fails because of `group3`
Something is making it spread over the blue rect.

File: src/shapes/group.class.js (hunk 9 above)
Comment (ShaMan123) [reply src/shapes/group.class.js]: Something is still wrong
image

Comment (ShaMan123) [reply src/shapes/group.class.js]: I have found the bug.
Give me a few minutes

Comment (ShaMan123) [reply src/shapes/group.class.js]: fixed 847284f7b93580ebe3b9fbe165bf209417ae009e

**Comment by ShaMan123** [comment]
@asturur I saw you made a lot a changes to group. Why?

**Comment by asturur** [comment]
I don't think i made changes willingly.
I may have tried to not bring in some parts that aren't related to the current group functionality and i may have changed some line trying to make visual test pass.

File: src/shapes/group.class.js [hunk 10]
```diff
@@ -383,8 +383,7 @@
      * Execute the drawing operation for an object on a specified context
      * @param {CanvasRenderingContext2D} ctx Context to render on
      */
-    drawObject: function (ctx) {
-      this._renderBackground(ctx);
```
Comment (ShaMan123) [inline src/shapes/group.class.js]: why don't you want this?
or you want it in the next PR?
it's instead of fill property.

Comment (asturur) [reply src/shapes/group.class.js]: group does not have fill today.
I'm trying to separate refactor from new features where possible.
I'm trying to have a series of (not so) small PRs where in case of errors, we can trace back where the behavour changed and why.

If this PR can have group with the new code with the listener objects, and all the other stuff working as. it was working before, we made a huge step forward

**Comment by asturur** [comment]
please don't commit on top, or i ll never get to the finish line. i know that line removed is breaking some visual test. i'll skip them.

File: src/shapes/group.class.js [hunk 11]
```diff
@@ -15,280 +21,318 @@
...
+          multiplyTransformMatrices(
+            invertTransform(this.calcTransformMatrix()),
+            object.calcTransformMatrix()
+          )
+        );
+      }
+      object.setCoords();
+      object._set('group', this);
       object._set('canvas', this.canvas);
+      this._watchObject(true, object);
```
Comment (ShaMan123) [inline src/shapes/group.class.js:255]:
```suggestion
      this.interactive && this._watchObject(true, object);
```

File: src/shapes/group.class.js [hunk 12]
```diff
@@ -15,280 +21,318 @@
...
+      var activeObject = this.canvas && this.canvas.getActiveObject && this.canvas.getActiveObject();
+      // if we are adding the activeObject in a group
+      if (activeObject && (activeObject === object || (object.contains && object.contains(activeObject)))) {
```
Comment (ShaMan123) [inline src/shapes/group.class.js:258]: why did you change this? bad perf

File: src/shapes/group.class.js [hunk 13]
```diff
@@ -370,151 +414,436 @@
...
+    prepareInitialBoundingBox: function (layoutDirective, objects, context) {
+      var options = context.options || {},
+          hasX = typeof options.left === 'number',
+          hasY = typeof options.top === 'number',
+          hasWidth = typeof options.width === 'number',
+          hasHeight = typeof options.height === 'number';
+
+      //  performance enhancement
+      //  skip layout calculation if bbox is defined
+      if ((hasWidth && hasHeight) || context.objectsRelativeToGroup || objects.length === 0) {
```
Comment (ShaMan123) [inline src/shapes/group.class.js:654]: why did you change this?

**Review by ShaMan123 — COMMENTED** [review]
@asturur it is extremely difficult for me to understand what you changed.
git diff is causing a nightmare.
I have left comments of places I notied a change that I disagree of or don't understand.
Plus I don't understand why you reintroduced the shadow logic. It is wrong and causes bugs. I verified that these shadow/group bugs were fixed once I removed the shadow methods. No need for them as far as I can tell.
Plus you can remove the `svg` layout directive logic

File: src/shapes/group.class.js (hunk 13 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:654]: This means that setting left/top values on group initialization will have no effect.
BUG

File: src/shapes/group.class.js [hunk 14]
```diff
@@ -15,280 +21,318 @@
...
+     * @private
+     * @param {boolean} watch
+     * @param {fabric.Object} object
+     */
+    _watchObject: function (watch, object) {
+      var directive = watch ? 'on' : 'off';
+      //  make sure we listen only once
+      watch && this._watchObject(false, object);
+      // TODO check what changed ischeck
+      object[directive]('changed', this.__objectMonitor);
```
Comment (ShaMan123) [inline src/shapes/group.class.js:221]: changed it a textbox event

File: src/shapes/group.class.js [hunk 15]
```diff
@@ -15,280 +21,318 @@
...
+    /**
+     * @private
+     * @param {boolean} watch
+     * @param {fabric.Object} object
+     */
+    _watchObject: function (watch, object) {
+      var directive = watch ? 'on' : 'off';
+      //  make sure we listen only once
+      watch && this._watchObject(false, object);
+      // TODO check what changed ischeck
```
Comment (ShaMan123) [inline src/shapes/group.class.js:220]:
```suggestion

```
changed is a textbox event

File: src/shapes/group.class.js (hunk 12 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:258]: not only is it bad perf but it is wrong as well because it checks only 1 level of nesting

File: src/shapes/active_selection.class.js (hunk 2 above)
Comment (asturur) [reply src/shapes/active_selection.class.js:40]: well once it did not need it. Now it seems to need it.
But apart from that that is minor, you may want to fire events for nested targets, but not need to select them.
We could argue that events ( mouse up, mouse down ) could fire 100% of the time since this is a breaking version, but i m not sure that would be an issue for some application that added on the mousedown/mouseup events logic that won't work on group. ( they can still filter it out checking if an object has group )

**Comment by ShaMan123** [comment]
updated comments

File: src/shapes/group.class.js (hunk 13 above)
Comment (asturur) [reply src/shapes/group.class.js:654]: i changed this because i was experimenting with the SVGs and this helped me fix it before i figured out the better option was not to pass in the svg options when grouping

File: src/shapes/group.class.js (hunk 12 above)
Comment (asturur) [reply src/shapes/group.class.js:258]: because i did not import the other function yet. the whole descendant of is not in yet

**Comment by asturur** [comment]
```
I have left comments of places I noticed a change that I disagree of or don't understand.
I don't understand why you reintroduced the shadow logic. It is wrong and causes bugs. I verified that these shadow/group bugs were fixed once I removed the shadow methods. No need for them as far as I can tell.
you can remove the svg layout directive logic
I don't understand why you removed the _renderObjects method
```

Because i wasn't done yet.
i m sure i left out rendering parts because are not related to the layout logic.
The renderObjects stuff is needed because of optimizations with the Group._activeObejects array
This PR contains only the fit-content layout, the improved geometric calculation and the watchers.

**Comment by asturur** [comment]
Ok now i can document any difference and we can argue if is part of the refactor of is a new feature that we can merge on its own.

**Comment by ShaMan123** [comment]
The last force push should be reverted as it reintroduced regressions

**Comment by ShaMan123** [comment]
I prefer to continue in a chat

File: src/shapes/group.class.js (hunk 10 above)
Comment (ShaMan123) [reply src/shapes/group.class.js]: Alright

File: src/shapes/active_selection.class.js (hunk 2 above)
Comment (ShaMan123) [reply src/shapes/active_selection.class.js:40]: Didn't understand

**Comment by ShaMan123** [comment]
closes #7811

**Comment by ShaMan123** [comment]
I think we should rebase #7761 on top of this branch and merge it here

File: src/util/misc.js (hunk 8 above)
Comment (ShaMan123) [reply src/util/misc.js]: I still see an difference in outputs margin but it doesn't surface in the visual tests

File: src/static_canvas.class.js [hunk 16]
```diff
@@ -558,6 +559,43 @@
...
+     * An object should be an instance of (or inherit from) fabric.Object
+     * @param {fabric.Object|fabric.Object[]} objects Object(s) to insert
+     * @param {Number} index Index to insert object at
+     * @param {Boolean} nonSplicing When `true`, no splicing (shifting) of objects occurs
+     * @return {Self} thisArg
+     * @chainable
+     */
+    insertAt: function (objects, index) {
+      fabric.Collection.insertAt.call(this, objects, index, this._onObjectAdded);
+      this.renderOnAddRemove && this.requestRenderAll();
```
Comment (ShaMan123) [inline src/static_canvas.class.js:584]:
```suggestion
      objects.length > 0 && this.renderOnAddRemove && this.requestRenderAll();
```

**Comment by asturur** [comment]
> I think we should rebase https://github.com/fabricjs/fabric.js/pull/7761 on top of this branch and merge it here

We can do that, i didn't have a chance to deeply study the layout strategy logic apart the early returns, so for me changing it now is the same as later.

Let me see if i m able to

**Comment by asturur** [comment]
i applied the PR opening yours, copying the entire function and overwriting the one i had, since rebase got nasty quickly for me. ( plus the other minor changes to types and to initialize ).
The branch does not look bad to me at allow, it has the functionalities of basic fabricJS.
I didn't try yet a bunch of things i wanted to try, but i ll do tomorrow mornin asap.

**Comment by ShaMan123** [comment]
Regarding shadow+group take a look at #7454

**Comment by ShaMan123** [comment]
What about #7781 ? can we add to here as well?

File: src/shapes/group.class.js (hunk 13 above)
Comment (ShaMan123) [reply src/shapes/group.class.js]: resolved 70bc1493a45c7d8e35d7e0fa407175249cac117c

File: src/shapes/group.class.js (hunk 12 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:259]: `deep` true as we discussed or simply import isDescendant of 
I would add a TODO comment so it's visible this is not complete.

**Comment by ShaMan123** [comment]
It seems multi selection is working as expected once `subTargetCheck=true`

**Comment by asturur** [comment]
yes the fact that it requires subTargetCheck set to true, is what i wanted to look into.

**Comment by asturur** [comment]
i see what it is, is a one line fix

File: src/shapes/active_selection.class.js [hunk 17]
```diff
@@ -100,6 +100,17 @@
       return false;
     },
 
+    /**
+     * @override
+     * @return {Boolean}
+     */
+    setCoords: function () {
```
Comment (ShaMan123) [inline src/shapes/active_selection.class.js:107]: isn't this redundant?
group has exactly this

Comment (asturur) [reply src/shapes/active_selection.class.js:107]: no it has it with subtargetcheck as an option indeed.

Comment (ShaMan123) [reply src/shapes/active_selection.class.js]: I see.

File: src/shapes/group.class.js (hunk 11 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:256]: 7d793b37

File: src/shapes/group.class.js (hunk 15 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:221]: 7d793b37

File: src/shapes/group.class.js (hunk 12 above)
Comment (ShaMan123) [reply src/shapes/group.class.js:259]: Added a TODO comment 7d793b37
